### PR TITLE
Add sequencing artifact manager

### PR DESCRIPTION
**Pull request by seallard**
This PR adds a class and utility functions for managing (retrieving and updating) artifacts from lims related to sequencing. This class will be used in the new script for the sample sequencing quality control. Part of project https://github.com/Clinical-Genomics/project-planning/issues/482. 

So this class will be a component of the new script replacing https://github.com/Clinical-Genomics/clinical_EPPs/blob/master/EPPs/bcl2fastq.py. 

I don't like how the new functions spread out across so many files, please let me know if there are other spots where they belong better or already exist. I don't think there was a super obvious structure to follow, so I bet that can be improved.

### Added
- Class for managing artifacts related to sequencing
- Utility function for managing artifacts related to sequencing

This [version](https://semver.org/) is a:
- [x] **MINOR** - when you add functionality in a backwards compatible manner

### CODE REVIEW

File: tests/EPPs/test_sequencing_artifact_manager.py [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+
+    # THEN the q30 threshold should have been set
+    assert isinstance(q30_threshold, int)
+    assert q30_threshold is not 0
+
+
+def test_sample_artifacts_initialization(
+    lims_process_with_novaseq_data: Process, lims: Lims
+):
+    """Test that the internal data structure holding the sample artifacts is populated."""
```
Comment (seallard) [inline tests/EPPs/test_sequencing_artifact_manager.py:42]: Not a fan of writing tests like this which breaks the encapsulation of the class. I want to test the behaviour, not the implementation. If the data structure for `sample_artifacts` ever change this test will break, it is brittle.

File: cg_lims/EPPs/qc/sequencing_artifact_manager.py [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+
+        if not sample_artifact:
+            LOG.warning(f"No metrics found for {sample_lims_id} in lane {lane}. Skipping.")
+            return
+
+        set_sample_reads(sample_artifact=sample_artifact, reads=reads)
+        set_sample_q30_score(sample_artifact=sample_artifact, q30_score=q30_score)
+        set_quality_check_flag(artifact=sample_artifact, quality_check_passed=passed_quality_control)
+
+        sample_artifact.put()
```
Comment (seallard) [inline cg_lims/EPPs/qc/sequencing_artifact_manager.py:79]: Maybe `put` should be called in the setters

File: tests/EPPs/test_sequencing_artifact_manager.py
Comment (seallard) [inline tests/EPPs/test_sequencing_artifact_manager.py:1]: All the tests are for the "happy path". Consider possible error execution paths and add tests for them

File: cg_lims/get/fields.py [hunk 3]
```diff
@@ -109,3 +109,19 @@ def get_barcode(artifact: Artifact):
         )
         return None
     
+def get_artifact_lims_id(sample_artifact: Artifact) -> Optional[str]:
+    samples = sample_artifact.samples if sample_artifact else None
+    if not (samples and samples[0].id):
+        return None
+    return samples[0].id
```
Comment (ChrOertlin) [inline cg_lims/get/fields.py:116]: this could be part of the class structure I mentioned above to replace the dict I suppose.

File: cg_lims/set/udfs.py [hunk 4]
```diff
@@ -8,6 +8,8 @@
 
 LOG = logging.getLogger(__name__)
 
+READS_FIELD = "# Reads"
+Q30_FIELD = "% Bases >=Q30"
```
Comment (ChrOertlin) [inline cg_lims/set/udfs.py:12]:
```suggestion
class UdfFields(strEnum)
READS_FIELD = "# Reads"
Q30_FIELD = "% Bases >=Q30"
```

are these used for udfs only? Maybe we can call them UdfConstants

File: tests/EPPs/test_sequencing_artifact_manager.py [hunk 5]
```diff
@@ -0,0 +1,94 @@
...
+        )
+
+    # THEN the sample artifacts should have been updated
+    for sample_id, lane in sample_lane_pairs:
+        sample_artifact = artifact_manager._get_sample_artifact(
+            sample_lims_id=sample_id, lane=lane
+        )
+        assert sample_artifact is not None
+        assert sample_artifact.udf[Q30_FIELD] == 0
+        assert sample_artifact.udf[READS_FIELD] == 0
```
Comment (ChrOertlin) [inline tests/EPPs/test_sequencing_artifact_manager.py:93]: With what value are these initialised? None or 0? Might be nice to add some number other than 0 and check for it. In mind comes some code in cg where we set things to 0 if None. Not sure how that is here

**Review by ChrOertlin — COMMENTED** [review]
Hi looks nice. Made some comments about potential structure changes that could make it more robust with the use of BaseModels instead of nested dicts. Would likely require some work. But up to you.

I think @Karl-Svard  has to approve changes here as the code owner, so commetning for now.

File: cg_lims/get/artifacts.py [hunk 6]
```diff
@@ -23,6 +25,13 @@ def get_sample_artifact(lims: Lims, sample: Sample) -> Artifact:
         raise MissingArtifactError(message=message)
     return artifact
 
+def get_sample_artifacts(lims: Lims, process: Process) -> List[Artifact]:
+    return get_output_artifacts(
+        lims=lims,
+        process=process,
+        output_type=ANALYTE_TYPE,
```
[the remark this answers is not in the document]
Comment (seallard) [reply cg_lims/get/artifacts.py:32]: Fixing this in a separate PR and will update the call here to use the enum https://github.com/Clinical-Genomics/cg_lims/pull/409

File: cg_lims/get/artifacts.py [hunk 7]
```diff
@@ -23,6 +25,13 @@ def get_sample_artifact(lims: Lims, sample: Sample) -> Artifact:
         raise MissingArtifactError(message=message)
     return artifact
 
+def get_sample_artifacts(lims: Lims, process: Process) -> List[Artifact]:
+    return get_output_artifacts(
```
[the remark this answers is not in the document]
Comment (seallard) [reply cg_lims/get/artifacts.py:29]: Well, me neither 😅  I'll talk to @Karl-Svard to get some clarification and maybe the doc string can be clarified. I think it is an implementation detail that the caller does not need to be aware of though

File: cg_lims/set/udfs.py (hunk 4 above)
Comment (seallard) [reply cg_lims/set/udfs.py:12]: Yeah, I think we should use an enum for these 👍

File: tests/EPPs/test_sequencing_artifact_manager.py (hunk 5 above)
Comment (seallard) [reply tests/EPPs/test_sequencing_artifact_manager.py:93]: Good point, maybe we can store the values before and check that they are not the same instead.

File: cg_lims/EPPs/files/sample_sheet/create_sample_sheet.py [hunk 8]
```diff
@@ -22,9 +22,12 @@
 LOG = logging.getLogger(__name__)
 
 
-def get_artifact_lane(artifact: Artifact) -> int:
+def get_artifact_lane(artifact: Artifact) -> Optional[int]:
     """Return the lane where an artifact is placed"""
-    return int(artifact.location[1].split(":")[0])
+    location = artifact.location if artifact else None
+    if not (location and location[1] and ":" in location[1]):
+        return None
```
[the remark this answers is not in the document]
Comment (ChrOertlin) [reply cg_lims/EPPs/files/sample_sheet/create_sample_sheet.py:29]: Yes, makes sense for this PR. Maybe refinement meeting can be held on cg_lims where some behaviour can be questioned. I guess we all know very little about this repo.

[the remark this answers is not in the document]
Comment (seallard) [reply cg_lims/EPPs/files/sample_sheet/create_sample_sheet.py]: Leaving it for a separate PR

File: cg_lims/EPPs/qc/sequencing_artifact_manager.py [hunk 9]
```diff
@@ -0,0 +1,79 @@
...
+
+LOG = logging.getLogger(__name__)
+
+
+class SequencingArtifactManager:
+    def __init__(self, process: Process, lims: Lims):
+        self.process: Process = process
+        self.lims: Lims = lims
+
+        self.sample_artifacts: Dict[str, Dict[int, Artifact]] = {}
```
[the remark this answers is not in the document]
Comment (seallard) [reply cg_lims/EPPs/qc/sequencing_artifact_manager.py]: Added a separate class containing the data structure and the add/get functions

[the remark this answers is not in the document]
Comment (seallard) [reply cg_lims/EPPs/qc/sequencing_artifact_manager.py]: Skipped using Pydantic, does not seem to fit this kind of data structure

File: cg_lims/get/artifacts.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (seallard) [reply cg_lims/get/artifacts.py]: Removed this function

File: cg_lims/get/fields.py (hunk 3 above)
Comment (seallard) [reply cg_lims/get/fields.py:116]: Leaving it for now, I don't think it belongs in the SampleArtifacts data class

File: tests/EPPs/test_sequencing_artifact_manager.py (hunk 5 above)
Comment (seallard) [reply tests/EPPs/test_sequencing_artifact_manager.py]: The fixture used in the test contains actual values for all these attributes

File: cg_lims/EPPs/files/sample_sheet/create_sample_sheet.py [hunk 10]
```diff
@@ -8,7 +8,7 @@
 from genologics.entities import Artifact, Process, ReagentType
 from cg_lims import options
 from cg_lims.exceptions import LimsError, InvalidValueError
-from cg_lims.get.artifacts import get_artifacts
+from cg_lims.get.artifacts import get_artifact_lane, get_artifacts
```
Comment (seallard) [inline cg_lims/EPPs/files/sample_sheet/create_sample_sheet.py:11]: I had to move `get_artifact_lane` to avoid a circular import

**Comment by Karl-Svard** [comment]
Nicely done! Just had some minor tweaks:

- Renamed some variables from `sample_artifact` to `sample_lane_artifacts` at several locations. This is to with "sample artifacts" being an already established term in cg_lims for a particular type of artifact that all samples get when registered in lims.
- Renamed `get_artifact_lims_id` to `get_artifact_sample_id` in order to better convey what it does.
- Renamed some of the `set` functions to make it more clear that it doesn't necessarily change the values on the sample level.
- Moved the test to match current folder structure.
- Some linting with black.

Please take look and just ask if you wonder anything!

File: cg_lims/get/artifacts.py [hunk 11]
```diff
@@ -22,6 +22,36 @@ class OutputGenerationType(str, Enum):
     PER_ALL_INPUTS = "PerAllInputs"
 
 
+def get_artifact_lane(artifact: Artifact) -> int:
+    """Return the lane where an artifact is placed"""
+    return int(artifact.location[1].split(":")[0])
```
Comment (seallard) [inline cg_lims/get/artifacts.py:27]: I think this function could use some error handling, but lets do it in a separate PR 👍

Comment (seallard) [reply cg_lims/get/artifacts.py:27]: Since it is existed from before and is used in other spots

File: cg_lims/get/artifacts.py [hunk 12]
```diff
@@ -22,6 +22,36 @@ class OutputGenerationType(str, Enum):
...
+
+    return list(lane_sample_artifacts)
+
+
+def is_output_type_per_reagent(output_map: Dict) -> bool:
+    return output_map["output-generation-type"] == OutputGenerationType.PER_REAGENT
+
+
+def get_artifact_from_map(map: Dict) -> Artifact:
+    return map["uri"]
```
Comment (seallard) [inline cg_lims/get/artifacts.py:52]: Magic string "uri", add to enum

**Review by Karl-Svard — APPROVED** [review]
Sweet! 💯
